lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,308 @@
+template <typename Decl_T>
+static bool isVisibleOutsideItsFile(const Decl_T *decl) {
+  return !hasSomePrivateAccessLevel(decl);
+}
 
-    default:
-      llvm_unreachable("everything else is a ValueDecl");
+class ReferenceDependencyFileEmitter {
+  SourceFile *SF;
+  const DependencyTracker &depTracker;
+  std::unique_ptr<llvm::raw_fd_ostream> out;
+
+  ReferenceDependencyFileEmitter(SourceFile *SF,
+                                 const DependencyTracker &depTracker,
+                                 std::unique_ptr<llvm::raw_fd_ostream> out)
+      : SF(SF), depTracker(depTracker), out(std::move(out)) {
+    assert(SF && "Cannot emit reference dependencies without a SourceFile");
+  }
+
+public:
+  static Optional<ReferenceDependencyFileEmitter>
+  create(DiagnosticEngine &diags, SourceFile *SF,
+         const DependencyTracker &depTracker, StringRef outputPath) {
+    // Before writing to the dependencies file path, preserve any previous file
+    // that may have been there. No error handling -- this is just a nicety, it
+    // doesn't matter if it fails.
+    llvm::sys::fs::rename(outputPath, outputPath + "~");
+
+    std::error_code EC;
+    auto out = llvm::make_unique<llvm::raw_fd_ostream>(outputPath, EC,
+                                                       llvm::sys::fs::F_None);
+
+    if (out->has_error() || EC) {
+      diags.diagnose(SourceLoc(), diag::error_opening_output, outputPath,
+                     EC.message());
+      out->clear_error();
+      return None;
     }
+
+    return ReferenceDependencyFileEmitter(SF, depTracker, std::move(out));
   }
 
-  return VD->getFormalAccess() <= AccessLevel::FilePrivate;
-}
+  void emit() {
+    *out << "### Swift dependencies file v0 ###\n";
 
-static bool extendedTypeIsPrivate(TypeLoc inheritedType) {
-  auto type = inheritedType.getType();
-  if (!type)
-    return true;
+    emitProviders();
+    emitDependencies();
+    emitInterfaceHash();
+    }
 
-  if (!type->isExistentialType()) {
-    // Be conservative. We don't know how to deal with other extended types.
-    return false;
+    void emitProviders() const;
+    void emitDependencies() const;
+
+    void emitInterfaceHash() {
+      llvm::SmallString<32> interfaceHashVal;
+      SF->getInterfaceHash(interfaceHashVal);
+      *out << interfaceHash << ": \"" << interfaceHashVal << "\"\n";
   }
+};
 
-  auto layout = type->getExistentialLayout();
-  assert(!layout.superclass && "Should not have a subclass existential "
-         "in the inheritance clause of an extension");
-  for (auto protoTy : layout.getProtocols()) {
-    if (!declIsPrivate(protoTy->getDecl()))
-      return false;
+class ProviderEmitter {
+  llvm::MapVector<const NominalTypeDecl *, bool> extendedNominals;
+  llvm::SmallVector<const FuncDecl *, 8> memberOperatorDecls;
+  llvm::SmallVector<const ExtensionDecl *, 8> extensionsWithJustMembers;
+
+  llvm::raw_fd_ostream &out;
+
+public:
+  ProviderEmitter(llvm::raw_fd_ostream &out) : out(out) {}
```

Comment:
Nitpick: `explicit`. (We really need a linter.)

Reply:
Sigh, yes, thanks! Fixed.